File: app/src/main/java/com/yp/payment/ui/LoginActivity.java
package com.yp.payment.ui;

import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Build;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

import com.yp.baselib.helper.NetListenHelper;
import com.yp.baselib.utils.NetUtils;
import com.yp.payment.Constant;
import com.yp.payment.Consts;
import com.yp.payment.R;
import com.yp.payment.dao.ShopConfigDao;
import com.yp.payment.http.MyCallback;
import com.yp.payment.internet.LoginRequest;
import com.yp.payment.internet.LoginResponse;
import com.yp.payment.internet.LoginResponseV2;
import com.yp.payment.internet.MyRetrofit;
import com.yp.payment.model.ShopConfig;
import com.yp.payment.utils.GsonUtil;

import kotlin.Unit;
import kotlin.jvm.functions.Function0;
import kotlin.jvm.functions.Function1;
import retrofit2.Call;

/**
 * @author : cp
 * @email : ibsfiq@qq.com
 * @date : 2019/8/21 14:07
 * @description ：管理员登录页
 */
public class LoginActivity extends BaseActivity {
    private static final String TAG = "LoginActivity";
    EditText edit_user_account, edit_user_psw;

    private static String[] PERMISSIONS_STORAGE = {Manifest.permission.READ_PHONE_STATE,
            Manifest.permission.READ_EXTERNAL_STORAGE,
            Manifest.permission.CAMERA,
            Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.CAMERA, Manifest.permission.ACCESS_COARSE_LOCATION};
    final private int REQUEST_PERMISSION_CODE = 20001;

    public static String deviceId;

    public ShopConfigDao shopConfigDao;
    private NetListenHelper netHelper;

    @Override
    public int layoutId() {
        return R.layout.activity_login;
    }

    @Override
    public void initView() {
        edit_user_account = findViewById(R.id.edit_user_account);
        edit_user_psw = findViewById(R.id.edit_user_psw);
        findViewById(R.id.btn_login_user).setOnClickListener(v->loginAdmin());

        netHelper = new NetListenHelper();
        netHelper.register(this,
                integer -> {
                    edit_user_account.setVisibility(View.VISIBLE);
                    edit_user_psw.setVisibility(View.VISIBLE);
                    return null;
                },
                () -> {
                    edit_user_account.setVisibility(View.INVISIBLE);
                    edit_user_psw.setVisibility(View.INVISIBLE);
                    ((Button)findViewById(R.id.btn_login_user)).setText("离线登录");
                    return null;
                }
                );

        checkBluetoothPermission();

        deviceId = android.os.Build.SERIAL;
    }

    void loginAdmin() {
        String user_account = edit_user_account.getText().toString().trim();
        String user_psw = edit_user_psw.getText().toString().trim();
        /*if (TextUtils.isEmpty(user_account)) {
            showToast(getResString(R.string.invalid_account));
            return;
        } else if (TextUtils.isEmpty(user_psw)) {
            showToast(getResString(R.string.invalid_psw));
            return;
        }*/

        final LoginRequest loginRequest = new LoginRequest();
        loginRequest.setDeviceId(LoginActivity.deviceId);
        loginRequest.setUsername(user_account);
        loginRequest.setPassword(user_psw);
//        loginRequest.setUsername("18312345678");
//        loginRequest.setPassword("000000");

        Log.d(TAG, "loginRequest==" + GsonUtil.GsonString(loginRequest));

        /*MyRetrofit.getApiService().init(loginRequest).enqueue(new MyCallback<LoginResponse>() {

            @Override
            public void onSuccess(LoginResponse loginResponse) {
                Log.d(TAG, "loginResponse==" + loginResponse);

                if (loginResponse.getCode() == 200) {
                    Constant.shopId = loginResponse.getData().getShopId();
                    Constant.cashierDeskId = loginResponse.getData().getCashierDeskId();
                    Constant.curUsername = loginRequest.getUsername();
                    Constant.shopName = loginResponse.getData().getShopName();

                    if (Constant.shopId == ) {
                    }

                    shopConfigDao.insertData(Constant.shopId, Constant.cashierDeskId,
                            Constant.shopName, loginRequest.getUsername());
                    openActivity(MoneyActivity.class);
                }
            }

            @Override
            public void onFailure(Call<LoginResponse> call, Throwable t) {

                Log.d(TAG, "loginResponse onFailure==" + t.getMessage());
                super.onFailure(call, t);
            }
        });*/

        MyRetrofit.getApiService().initV2(loginRequest).enqueue(new MyCallback<LoginResponseV2>() {

            @Override
            public void onSuccess(LoginResponseV2 loginResponse) {
                Log.d(TAG, "loginResponse==" + loginResponse);

                if (loginResponse.getCode() == 200) {
                    Constant.shopId = loginResponse.getData().getShopId();
                    Constant.cashierDeskId = loginResponse.getData().getCashierDeskId();
                    Constant.curUsername = loginRequest.getUsername();
                    Constant.shopName = loginResponse.getData().getShopName();

                    if (loginResponse.getData().getCashAllow() != null &&
                            loginResponse.getData().getCashAllow().intValue() == 1) {
                        Consts.payModes = Consts.payModesWithCash;
                        Consts.payModeIcons = Consts.payModeIconsWithCash;
                    }else {
                        Consts.payModes = Consts.payModesWithNoCash;
                        Consts.payModeIcons = Consts.payModeIconsWithNoCash;
                    }

                    shopConfigDao.insertData(Constant.shopId, Constant.cashierDeskId,
                            Constant.shopName, loginRequest.getUsername());
                    openActivity(MoneyActivity.class);
                }
            }

            @Override
            public void onFailure(Call<LoginResponseV2> call, Throwable t) {
                Log.d(TAG, "loginResponse onFailure==" + t.getMessage());
//                showToast("网络异常");
                super.onFailure(call, t);
            }
        });

//        finish();


    }

    @Override
    public void initData() {
        shopConfigDao = new ShopConfigDao(this);

        ShopConfig shopConfig = shopConfigDao.query();

        if (shopConfig != null && shopConfig.getAutoLogin() == 1 && NetUtils.INSTANCE.isNetConnected(this)) {
            Constant.shopId = shopConfig.getShopId();
            Constant.cashierDeskId = shopConfig.getCashierDeskId();
            Constant.curUsername = shopConfig.getUsername();
            Constant.shopName = shopConfig.getShopName();
            Consts.payModes = Consts.payModesWithNoCash;
            Consts.payModeIcons = Consts.payModeIconsWithNoCash;
            openActivity(MoneyActivity.class);
        }
    }

    /*
   校验蓝牙权限
  */
    private void checkBluetoothPermission() {

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (ActivityCompat.checkSelfPermission(this, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this, PERMISSIONS_STORAGE, REQUEST_PERMISSION_CODE);
            } else if (ActivityCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this, PERMISSIONS_STORAGE, REQUEST_PERMISSION_CODE);
            } else if (ActivityCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this, PERMISSIONS_STORAGE, REQUEST_PERMISSION_CODE);
            } else if (ActivityCompat.checkSelfPermission(this, Manifest.permission.READ_PHONE_STATE) != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this, PERMISSIONS_STORAGE, REQUEST_PERMISSION_CODE);
            } else if (ContextCompat.checkSelfPermission(LoginActivity.this,
                    Manifest.permission.ACCESS_COARSE_LOCATION)
                    != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this, PERMISSIONS_STORAGE, REQUEST_PERMISSION_CODE);
            } else {
                Log.d(TAG, "无需授权  ");
            }
        } else {//小于23版本直接使用
            Log.d(TAG, "checkBluetoothPermission: 小于23版本直接使用");
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        netHelper.unregister(this);
    }
}

File: app/src/main/java/com/yp/payment/ui/LoginActivity.kt
package com.yp.payment.ui

import android.Manifest
import android.app.ProgressDialog
import android.os.Build
import android.os.Bundle
import android.util.Log
import com.yp.baselib.annotation.LayoutId
import com.yp.baselib.base.BaseActivity
import com.yp.baselib.helper.NetListenHelper
import com.yp.baselib.utils.NetUtils.isNetConnected
import com.yp.baselib.utils.PermissionUtils
import com.yp.payment.Constant
import com.yp.payment.Consts
import com.yp.payment.MainActivity
import com.yp.payment.R
import com.yp.payment.dao.ShopConfigDao
import com.yp.payment.http.MyCallback
import com.yp.payment.internet.LoginRequest
import com.yp.payment.internet.LoginResponseV2
import com.yp.payment.internet.MyRetrofit
import com.yp.payment.update.SPHelper
import com.yp.payment.utils.GsonUtil
import kotlinx.android.synthetic.main.activity_login.*
import retrofit2.Call

/**
 * @author : cp
 * @email : ibsfiq@qq.com
 * @date : 2019/8/21 14:07
 * @description ：管理员登录页
 */
@LayoutId(R.layout.activity_login)
class LoginActivity : BaseActivity() {
    private val REQUEST_PERMISSION_CODE = 20001
    var shopConfigDao: ShopConfigDao? = null
    private var netHelper: NetListenHelper? = null


    fun loginAdmin() {
        val user_account = edit_user_account!!.text.toString().trim { it <= ' ' }
        val user_psw = edit_user_psw!!.text.toString().trim { it <= ' ' }
        val loginRequest = LoginRequest()
        loginRequest.deviceId = deviceId
        loginRequest.username = user_account
        loginRequest.password = user_psw
        Log.d(TAG, "loginRequest==" + GsonUtil.GsonString(loginRequest))
//        Req.login(this, user_account, user_psw, {
//            shopConfigDao!!.insertData(SPHelper.getShopId(), SPHelper.getCashierDeskId(),
//                    SPHelper.getShopName(), SPHelper.getCurUsername())
//            goTo<MoneyActivity>()
//        }, {
//            "失败".toast()
//        })
        val pd = ProgressDialog(this)
        pd.setMessage("正在登录...")
        pd.show()
        MyRetrofit.getApiService().initV2(loginRequest).enqueue(object : MyCallback<LoginResponseV2?>() {
            override fun onSuccess(loginResponse: LoginResponseV2?) {
                pd.dismiss()
                Log.d(TAG, "loginResponse==$loginResponse")
                if (loginResponse?.code == 200) {
                    SPHelper.putShopId(loginResponse.data.shopId)
                    SPHelper.putCashierDeskId(loginResponse.data.cashierDeskId)
                    SPHelper.putCurUsername(loginRequest.username)
                    SPHelper.putShopName(loginResponse.data.shopName)

                    Constant.shopId = loginResponse.data.shopId
                    Constant.cashierDeskId = loginResponse.data.cashierDeskId
                    Constant.curUsername = loginRequest.username
                    Constant.shopName = loginResponse.data.shopName

                    if (loginResponse.data.cashAllow != null &&
                            loginResponse.data.cashAllow.toInt() == 1) {
                        Consts.payModes = Consts.payModesWithCash
                        Consts.payModeIcons = Consts.payModeIconsWithCash
                    } else {
                        Consts.payModes = Consts.payModesWithNoCash
                        Consts.payModeIcons = Consts.payModeIconsWithNoCash
                    }

//                    ObjectBox.boxStore.boxFor(ShopConfig::class.java)?.let {
//                        it.put(ShopConfig().apply {
//                            shopId = loginResponse.data.shopId
//                            cashierDeskId = loginResponse.data.cashierDeskId
//                            shopName = loginResponse.data.shopName
//                            username = loginRequest.username
//                        })
//                        it.all.forEach {
//                            it.toString().logD("TEST123")
//                        }
//                    }

                    shopConfigDao!!.insertData(loginResponse.data.shopId, loginResponse.data.cashierDeskId,
                            loginResponse.data.shopName, loginRequest.username)
                    goTo<MoneyActivity>()
                }
            }

            override fun onFailure(call: Call<LoginResponseV2?>, t: Throwable) {
                super.onFailure(call, t)
                pd.dismiss()
                t.localizedMessage.toast()
                Log.d(TAG, "loginResponse onFailure==" + t.message)
            }
        })
    }

    fun initData() {
        shopConfigDao = ShopConfigDao(this)
        val shopConfig = shopConfigDao!!.query()
        if (shopConfig != null && shopConfig.autoLogin == 1 && isNetConnected(this)) {
            Consts.payModes = Consts.payModesWithNoCash
            Consts.payModeIcons = Consts.payModeIconsWithNoCash
            goTo<MoneyActivity>(true)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        netHelper!!.unregister(this)
    }

    override fun init(bundle: Bundle?) {
        netHelper = NetListenHelper()
        netHelper!!.register(this,
                {
                    show(edit_user_account, edit_user_psw)
                    btn_login_user.text = "登录"
                }
        ) {
            hide(edit_user_account, edit_user_psw)
            btn_login_user.text = "离线登录"
        }
        deviceId = Build.SERIAL
        initData()
        btn_login_user.click { loginAdmin() }
        PermissionUtils.req(this, null, null, null,
                *PERMISSIONS_STORAGE)

        tv_ip.click {
            goTo<IpSetActivity>()
        }

    }

    companion object {
        private const val TAG = "LoginActivity"
        private val PERMISSIONS_STORAGE = arrayOf(
                Manifest.permission.READ_PHONE_STATE,
                Manifest.permission.READ_EXTERNAL_STORAGE,
                Manifest.permission.CAMERA,
                Manifest.permission.WRITE_EXTERNAL_STORAGE,
                Manifest.permission.ACCESS_COARSE_LOCATION)

        @JvmField
        var deviceId: String? = null
    }
}